ufl/indexed.py

```diff
@@ -41,9 +43,12 @@ def __new__(cls, expression, multiindex):
             else:
                 fi, fid = (), ()
             return Zero(shape=(), free_indices=fi, index_dimensions=fid)
-        elif expression.ufl_shape == () and multiindex == ():
-            return expression
-        else:
+
+        try:
+            # Simplify indexed ListTensor
+            c = expression[multiindex]
+            return Indexed(*c.ufl_operands) if isinstance(c, Indexed) else c
```

Comment:
Can we not do:
```suggestion
            return expression[multiindex]
```
?

Reply:
If `__new__` returns an old `Indexed` object this causes an infinite recursion. Rebuilding the object seems to avoid this issue.

Reply:
According to the [python documentation](https://docs.python.org/3/reference/datamodel.html#basic-customization):
> If `__new__()` returns an instance of cls, then the new instance’s `__init__()` method will be invoked like `__init__(self[, ...])`, where self is the new instance and the remaining arguments are the same as were passed to `__new__()`.

Reply:
Another way to prevent `Indexed.__init__()` to be called on already initialized instances is simply by leaving the `Indexed.__init__()` method unimplemented, and moving its contents to `Indexed.__new__()`.

Reply:
Is the `__new__() / __init__()` business the cause of infinite recursion? I can see that infinite recursion could be possible, though. Do you have an example which would cause infinite recursion to help me understand?

Reply:
That's right. You can use the code you are suggesting on this firedrake test to replicate the infinite recursion: `tests/firedrake/equation_bcs/test_equation_bcs.py::test_EquationBC_mixedpoisson_matrix[nonlinear]`

Reply:
I resolved this by introducing an `_init()` method called only on new instances within `__new__()`, in the same way other `Operator` subclasses like `Sum` do this.